app/grandchallenge/cases/widgets.py

```diff
@@ -0,0 +1,114 @@
+from django.core.exceptions import ValidationError
+from django.forms import (
+    HiddenInput,
+    ModelChoiceField,
+    ModelMultipleChoiceField,
+    MultiValueField,
+    MultiWidget,
+)
+from django.forms.widgets import ChoiceWidget
+from django.utils.datastructures import MultiValueDictKeyError
+
+from grandchallenge.cases.models import Image
+from grandchallenge.cases.utils import WidgetChoices
+from grandchallenge.uploads.widgets import UserUploadMultipleWidget
+
+
+class ImageSearchMixin(ChoiceWidget):
+    template_name = "cases/image-search-widget.html"
+    input_type = None
+    name = None
+
+    def __init__(self, *args, name=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        if name:
+            self.name = name
+
+    def get_context(self, *args, **kwargs):
+        context = super().get_context(*args, **kwargs)
+        if self.name:
+            context["widget"]["name"] = self.name
+        return context
+
+
+class ImageSearchWidget(ImageSearchMixin, HiddenInput):
+    pass
+
+
+class FlexibleImageWidget(MultiWidget):
+    template_name = "cases/flexible_image_widget.html"
+
+    def __init__(
+        self,
+        *args,
+        help_text=None,
+        user=None,
+        current_value=None,
+        **kwargs,
+    ):
+        widgets = (
+            ImageSearchWidget(),
+            UserUploadMultipleWidget(),
+        )
+        super().__init__(widgets)
+        self.attrs = {
+            "help_text": help_text,
+            "user": user,
+            "current_value": current_value,
+            "widget_choices": {
+                choice.name: choice.value for choice in WidgetChoices
+            },
+        }
+
+    def decompress(self, value):
+        if value:
+            if Image.objects.filter(pk=value).exists():
+                return [value, None]
+            else:
+                return [None, [value]]
+        else:
+            return [None, None]
+
+    def value_from_datadict(self, data, files, name):
+        try:
+            value = data[name]
+        except MultiValueDictKeyError:
+            try:
+                # this happens if the data comes from the DS update form
+                value = data[f"WidgetChoice-{name}"]
+            except MultiValueDictKeyError:
+                value = None
+        if value:
+            if Image.objects.filter(pk=value).exists():
+                return [value, None]
+            else:
+                return [None, [value]]

```

Comment:
Are these covered by tests because TBH I have no idea what is happening here.

Reply:
Yes, although I see some room for improvements. Will make things a bit more explicit